src/brain/index_selection.cpp

```diff
@@ -0,0 +1,449 @@
+  // Get the cheapest indexes through exhaustive search upto a threshold
+  ExhaustiveEnumeration(indexes, top_indexes, workload);
+
+  // Get all the remaining indexes which can be part of our optimal set
+  auto remaining_indexes = indexes - top_indexes;
+
+  // Greedily add the remaining indexes until there is no improvement in the
+  // cost or our required size is reached
+  GreedySearch(top_indexes, remaining_indexes, workload, num_indexes);
+}
+
+void IndexSelection::GreedySearch(IndexConfiguration &indexes,
+                                  IndexConfiguration &remaining_indexes,
+                                  Workload &workload, size_t k) {
+  // Algorithm:
+  // 1. Let S = the best m index configuration using the naive enumeration
+  // algorithm. If m = k then exit.
+  // 2. Pick a new index I such that Cost (S U {I}, W) <= Cost(S U {I'}, W) for
+  // any choice of I' != I
+  // 3. If Cost (S U {I}) >= Cost(S) then exit
+  // Else S = S U {I}
+  // 4. If |S| = k then exit
+
+  size_t current_index_count = context_.naive_enumeration_threshold_;
+
+  if (current_index_count >= k) return;
+
+  double global_min_cost = ComputeCost(indexes, workload);
+  double cur_min_cost = global_min_cost;
+  double cur_cost;
+  std::shared_ptr<IndexObject> best_index;
+
+  // go through till you get top k indexes
+  while (current_index_count < k) {
+    // this is the set S so far
+    auto original_indexes = indexes;
+    for (auto index : remaining_indexes.GetIndexes()) {
+      indexes = original_indexes;
+      indexes.AddIndexObject(index);
+      cur_cost = ComputeCost(indexes, workload);
+      if (cur_cost < cur_min_cost) {
+        cur_min_cost = cur_cost;
+        best_index = index;
+      }
+    }
+
+    // if we found a better configuration
+    if (cur_min_cost < global_min_cost) {
+      indexes.AddIndexObject(best_index);
+      remaining_indexes.RemoveIndexObject(best_index);
+      current_index_count++;
+      global_min_cost = cur_min_cost;
+
+      // we are done with all remaining indexes
+      if (remaining_indexes.GetIndexCount() == 0) {
+        break;
+      }
+    } else {  // we did not find any better index to add to our current
+              // configuration
+      break;
+    }
+  }
+}
+
+void IndexSelection::ExhaustiveEnumeration(IndexConfiguration &indexes,
+                                           IndexConfiguration &top_indexes,
+                                           Workload &workload) {
+  // Get the best m index configurations using the naive enumeration algorithm
+  // The naive algorithm gets all the possible subsets of size <= m and then
+  // returns the cheapest m indexes
+  assert(context_.naive_enumeration_threshold_ <= indexes.GetIndexCount());
+
+  // Define a set ordering of (index config, cost) and define the ordering in
+  // the set
+  std::set<std::pair<IndexConfiguration, double>, IndexConfigComparator>
+      running_index_config(workload);
```

Comment:
nit: declare those together to avoid writing the long typename over and over again

Reply:
Done